common/transport/amqp/src/amqp.ts

```diff
@@ -145,44 +235,190 @@ export class Amqp {
+            // will occur until the nextTick() after the call to create_session.  Because of that, one can
+            // put the event handlers on the rhea session object returned from the create_session call and be assured
+            // that the listeners are in place BEFORE any possible events will be emitted on the session.
+            //
+            this._rheaSession = this._rheaConnection.create_session();
+            manageSessionHandlers('on');
+            this._rheaSession.open();
+          },
+          'session_open': (context: EventContext) => {
+            debug('In the session_open handler for the connecting_session state');
+            let callback = this._sessionCallback;
+            let result = this._sessionResult;
+            this._sessionCallback = undefined;
+            this._sessionResult = undefined;
+            this._fsm.transition('connected', callback, result);
+          },
+          'session_error': (context: EventContext) => {
+            debug('In the session_error handler for the connecting_session state with error of: ' + this._getErrorName(context.session.error));
+            debug('Deferring transition until the session_close event occurs.');
+            this._indicatedSessionError = context.session.error;
+          },
+          'session_close': (context: EventContext) => {
+            debug('In the session_close handler for the connecting_session state with indicated error of: ' + this._getErrorName(this._indicatedSessionError));
+            debug('Transition to disconnecting');
+            let err = this._indicatedSessionError;
+            let callback = this._sessionCallback;
+            this._indicatedSessionError = undefined;
+            this._sessionCallback = undefined;
+            this._sessionCloseOccurred = true;
+            this._fsm.transition('disconnecting', callback, err);
+          },
+          'connection_error': (context: EventContext) => {
+            debug('In the connection_error handler for the connecting session state with error of: ' + this._getErrorName(context.connection.error));
+            debug('Deferring transition until the connection_close occurs.');
+            this._indicatedConnectionError = context.connection.error;
+          },
+          'connection_close': (context: EventContext) => {
+            debug('In the connection_close handler for the connecting_session state with indicated error of: ' + this._getErrorName(this._indicatedConnectionError));
+            debug('Transition to disconnecting');
+            let err = this._indicatedConnectionError;
+            let callback = this._sessionCallback;
+            //
+            // We lie about session close coming in.  Thing is that if we are here we don't actually have
+            // a good session set up.  This way we won't wait around for a session end that probably won't come.
+            //
```

Comment:
this is the "global state" anti-pattern coming back to bite you. you now have to manage consistency of that state which is a collection of disjointed variables.

Reply:
I disagree.  We are dealing with a fairly bizarre fever dream of error conditions.  But we deal with it.  Run downs are always hard.

---
In reply to: [209132891](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209132891) [](ancestors = 209132891)